components/ee/ws-scheduler/pkg/scheduler/scheduler.go

```diff
@@ -609,6 +652,25 @@ func (s *Scheduler) isPendingPodWeHaveToSchedule(pod *corev1.Pod) bool {
 		pod.ObjectMeta.Namespace == s.Config.Namespace
 }
 
+// deleteGhostWorkspace tries to delete a ghost workspace to make room for the given pod
+func (s *Scheduler) deleteGhostWorkspace(ctx context.Context, podName string, ghostName string) (err error) {
+	span, ctx := tracing.FromContext(ctx, "deleteGhostWorkspace")
+	defer tracing.FinishSpan(span, &err)
+
+	err = s.Clientset.CoreV1().Pods(s.Config.Namespace).Delete(ghostName, metav1.NewDeleteOptions(1))
```

Comment:
TODO: is using `gracePeriod=0` instead viable? e.g., does it leave zombie process behind?

Reply:
Yes it would. grace-period == 0 deletes instantly from the data plane and does not wait at all for the process to actually stop (looking at the kubectl sources reveals that `--force` just guards `--grace-period=0`, but it's the latter that actually forces removal). We would run the risk of leaking resources on the node.

Reply:
Would be nice if we found a way to shave off this const 1 second offset on (nearly) every workspace start...

Reply:
That one second is just the grace period, things can go quicker - but also slower.
The main downside of this approach is that we incur the "kubernetes control plane latency" several times.

We could just force delete and have a GC on the node, akin to the [containerd 4124 fix](https://github.com/gitpod-io/gitpod/blob/master/components/ws-daemon/pkg/daemon/containerd4214.go). Only thing I'm not sure is how the kubelet does the accounting, i.e. if we run the risk of a workspace failing on the node because the kubelet still sees the pod, even though it's gone from the control plane.

Reply:
> We could just force delete and have a GC on the node
That would be ideal in terms of startup perf, but also feels quite "heavy". I'd say let's see if this is a real issue or not.